cli/command/run/cmd.go

```diff
@@ -103,11 +111,57 @@ func newRunner(p config.Project, cli *command.SauceCtlCli, seq fleet.Sequencer)
 		if err != nil {
 			return nil, err
 		}
+
+		cip := createCIProvider()
+		seq := createCISequencer(p, cip)
+
 		log.Info().Msg("Starting CI runner")
-		return ci.NewRunner(p, cli, seq, rc)
+		return ci.NewRunner(p, cli, seq, rc, cip)
 	}
 	log.Info().Msg("Starting local runner")
-	return docker.NewRunner(p, cli, seq)
+	return docker.NewRunner(p, cli, &memseq.Sequencer{})
+}
+
+func createCIProvider() ci.Provider {
+	enableCIProviders()
+	cip := ci.Detect()
+	// Allow users to override the CI build ID
+	if ciBuildID != "" {
+		log.Info().Str("id", ciBuildID).Msg("Using user provided build ID.")
+		cip.SetBuildID(ciBuildID)
+	}
+
+	return cip
+}
+
+func createCISequencer(p config.Project, cip ci.Provider) fleet.Sequencer {
+	if !p.Parallel {
+		log.Info().Msg("Parallel execution is turned off. Running tests sequentially.")
+		return &memseq.Sequencer{}
+	}
+	u := os.Getenv("SAUCE_USERNAME")
+	k := os.Getenv("SAUCE_ACCESS_KEY")
+	if u == "" || k == "" {
+		log.Info().Msg("No credentials provided. Running tests sequentially.")
+		return &memseq.Sequencer{}
+	}
+	if cip == ci.NoProvider && ciBuildID == "" {
```

Comment:
@alexplischke should this be an OR statement?
```
if cip == ci.NoProvider || ciBuildID == ""
```

Reply:
I think I'll add a comment in the code to make that a bit more clear (actually, the README contains some explanation now as well regarding this check).

TLDR:
1. Check if we can detect a supported CI Provider. If not, we'll be left with `ci.NoProvider` to indicate that there isn't one.
2. Without a CI Provider, we can't generate a reliable CI build ID that is necessary for parallelization, so the user has to provide one.
3. (optional) The user is generally also free to provide their own CI build ID regardless of the detected CI Provider, in which case we'll just use what the user has given us.

Reply:
Got it, definitely an AND statement